Cadence Account Storage Map Migration

turbolent opened the pull request:
Implement the migration logic for onflow/cadence#3584

- Add a new system contract `AccountV2Migration`, which is deployed to the service account during bootstrapping. Its `Admin` resource is able to trigger the migration for the next batch of accounts to be migrated, by calling the injected `scheduleAccountV2Migration` function
- Expose a new function `fun scheduleAccountV2Migration(addressStartIndex: UInt64, count: UInt64): Bool` to the new contract `AccountV2Migration` in the service account. It generates the addresses in the given range, and schedules the migration to the new account storage format V2. It essentially just calls the internal Go function `runtime.Storage.ScheduleV2Migration()` provided by Cadence.

fxamacker commented:
@turbolent Thanks for working on this part of the integration!

Do we want to create a feature branch and target this PR to the feature branch?

turbolent commented:
@onflow/flow-cadence-execution is there a way to call `accountV2Migration.DeclareScheduleAccountV2MigrationFunction` during bootstrapping?

janezpodhostnik commented:
@turbolent during bootstrapping you have access to the function `invokeMetaTransaction` to that function you have to pass an fvm context. 

When you want to run `DeclareScheduleAccountV2MigrationFunction` you can create a child context `NewContextFromParent` to which you add `WithReusableCadenceRuntimePool` and add the new cadence runtime pool.

Let me know if that helps.

turbolent commented:
@janezpodhostnik Thanks! It turned out that setting up the functions in the runtime won't work, because an environments chain ID might change (default config chain is MN, but then actual chain is e.g. Emulator). I refactored the code to be similar to the EVM setup, i.e. declaring the functions in the transaction invoker and in the script invoker.

turbolent commented:
@fxamacker I fixed a few things and added a test in 8b597cbd1036bcef59f92fb51d8eb3d086cc6e48 which demonstrates that the migration of the service account works. Next step is probably to write an integration test which does this using the system chunk transaction, but I'm very new to this. Maybe someone from @onflow/flow-cadence-execution can take this on or at least point me in the right direction?

fxamacker commented:
@turbolent Thanks for adding the test and other updates!

Would you like me to start reviewing this PR or should I wait until integration test is added and DRAFT status is removed?  No rush, just checking before I begin something needed next week.

turbolent commented:
@fxamacker Feel free to already start reviewing it. We'll probably want to have the integration tests before we merge, but feedback is already appreciated, thanks!

fxamacker commented on fvm/accountV2Migration/AccountV2Migration.cdc:
```diff
@@ -0,0 +1,102 @@
+            AccountV2Migration.nextAddressStartIndex = nextAddressStartIndex
+        }
+
+        access(all)
+        fun setBatchSize(_ batchSize: UInt64) {
+            AccountV2Migration.batchSize = batchSize
+        }
```
Do we want to have a minimum `batchSize` that is larger than 0?  

It seems migration is effectively disabled when `batchSize` is 0.  Maybe forcing caller to explicitly use `setEnabled(false)` would reduce mistakes.

turbolent replied:
Good point, the enabled flag isn't really needed, we can disable via the batch size. Changed it in [f86dd29](https://github.com/onflow/flow-go/pull/6761/commits/f86dd292a2b788d198b31c73c7784ed9ca74ae56)

fxamacker reviewed (commented):
Nice! :+1:  I mostly reviewed from Go programming perspective, and left one minor comment.

turbolent commented:
As [discussed in the last Execution Team meeting](https://www.notion.so/flowfoundation/Execution-Team-meeting-agenda-82cc3be512df4497a1a7ce57cc786a40?pvs=4#1591aee1232480aba327eab165b59823), I've added another `Migration` system account contract, and changed the system chunk transaction to call into it instead. The `Migration` contract then calls into the `AccountV2Migration`. 

We'll have to deploy this PR in an HCU, given the system chunk transaction change, but with this change, we can repurpose the `Migration` logic in the future for other "on-the-fly" migrations, potentially saving an HCU.

janezpodhostnik commented on fvm/blueprints/scripts/systemChunkTransactionTemplate.cdc:
```diff
@@ -19,9 +20,12 @@ transaction {
-        }
+        let evmHeartbeat = serviceAccount.storage
+            .borrow<&EVM.Heartbeat>(from: /storage/EVMHeartbeat)
+        evmHeartbeat?.heartbeat()
```
I spent way to long looking for a `'.` operator in the cadence language docs only to find out there is something wrong with the fonts:
![image](https://github.com/user-attachments/assets/576e9c44-7b50-4298-9b0b-e2f29cd75881)

turbolent replied:
Where is this screenshot from? GitHub? VS Code? It's probably not a problem with the PR though?

janezpodhostnik replied:
This is github on firefox. No not an issue in the PR. Just some sort of font issue somewhere.

janezpodhostnik commented on fvm/accountV2Migration/AccountV2Migration.cdc:
```diff
@@ -0,0 +1,89 @@
+    var batchSize: UInt64
+
+    init() {
+        self.adminStoragePath = /storage/accountV2MigrationAdmin
+        self.nextAddressStartIndex = 1
+        self.batchSize = 0
+
+        self.account.storage.save(
+            <-create Admin(),
+            to: self.adminStoragePath
+        )
```
This is really small, but the batching logic could probably live in the `Migration` contract instead of in the V2 one

turbolent replied:
See https://github.com/onflow/flow-go/pull/6761#issuecomment-2549713980: The idea is that all logic related to the particular migration, the "account v2 migration", lives in `AccountV2Migration`, and `Migration` only delegates. 

If we add the logic of the "account v2 migration" to the `Migration` contract, we won't be able to reuse it for other purposes, as the contract update rules e.g. don't allow removing enums, etc.

janezpodhostnik reviewed (commented):
I imagined that `scheduleAccountV2Migration` would be on the `Migration` contract and would just be called `scheduleAccountMigration` which would then dispatch this to the V2 contract.

I also though we could avoid using the V2 contract directly anywhere in the FVM (excluding tests of course).

janezpodhostnik commented:
> Agreed, we should not use it directly, only through the Migration contract. Maybe I missed a spot?

The thing that threw me off was the reference to `accountV2Migration.DeclareFunctions` in the transaction invoker. 

Would it be possible to have `scheduleAccountMigration` on the `Migration` contract that then gets forwarded to `scheduleAccountV2Migration`?

Yes maybe lets just quickly sync.

janezpodhostnik commented:
Lets just bootstrap with `AccountV2Migration` for now. There is a separate issue to boot with a minimal set of contracts, and we can clean this up there.

fxamacker commented on fvm/accountV2Migration/AccountV2Migration.cdc:
```diff
@@ -0,0 +1,89 @@
+    access(all)
+    let adminStoragePath: StoragePath
+
+    access(all)
+    var nextAddressStartIndex: UInt64
+
+    access(all)
+    var batchSize: UInt64
```
If it is feasible, maybe we can have a `maxAddressIndex` corresponding to address index when we start migrating accounts. 

With something like a `maxAddressIndex`, batch migration can be disabled automatically.

turbolent replied:
Good idea! Added in [336e2dc](https://github.com/onflow/flow-go/pull/6761/commits/336e2dc0d351ad015d87169acbd96cbf828abde6)

fxamacker reviewed (approved):
Nice!  :+1: I left one minor comment about possibility of automatically disabling batch migration by adding and using something like a max address index.

janezpodhostnik commented on integration/internal/emulator/blockchain.go:
```diff
@@ -1034,6 +1045,7 @@ func (b *Blockchain) executeSystemChunkTransaction() error {
 		fvm.WithSequenceNumberCheckAndIncrementEnabled(false),
 		fvm.WithRandomSourceHistoryCallAllowed(true),
 		fvm.WithBlockHeader(b.pendingBlock.Block().Header),
+		fvm.WithAccountStorageLimit(false),
```
@turbolent I added this for feature parity.

turbolent replied:
Great, thank you!
